flang/lib/Optimizer/OpenMP/DoConcurrentConversion.cpp

```diff
@@ -366,24 +374,78 @@ void collectIndirectConstOpChain(mlir::Operation *link,
   opChain.insert(link);
 }
 
+/// Loop \p innerLoop is considered perfectly-nested inside \p outerLoop iff
+/// there are no operations in \p outerloop's other than:
+///
+/// 1. the operations needed to assing/update \p outerLoop's induction variable.
+/// 2. \p innerLoop itself.
+///
```

Comment:
I think we also need to test:
1. Computing the inner DoLoop's lb/ub/step does not depend on the outer loop's indvar
2. No instructions that have side-effects

[idea] It would be possible to just test whether the instructions can be sunk into the inner loop: it is not really relevant whether those instructions depend on the outer loop's indvar.

Reply:
> Computing the inner DoLoop's lb/ub/step does not depend on the outer loop's indvar

For loop nests (i.e. multi-range `do concurrent` loops) this is prohibited by the spec:
```
C1125 (R1126) A concurrent-limit or concurrent-step in a concurrent-control shall not contain a reference to
any index-name in the concurrent-control-list in which it appears.
```

> No instructions that have side-effects

You mean sibling instructions to the inner loop (i.e. either before or after it within the outer loop)? Or instruction within the body of the innermost loop in the nest?